tests/e2e/s3/utils.py

```diff
@@ -0,0 +1,269 @@
+
+    Args:
+        local_endpoint (str): The endpoint URL of the local S3 service.
+
+    Raises:
+        Exception: If the file is not found in the S3 bucket.
+
+    Returns:
+        None
+    """
+    s3 = boto3.client("s3", endpoint_url=local_endpoint)
+    found = False
+    for i in range(100):
+        response = s3.list_objects_v2(Bucket="metrics")
+        if "Contents" in response:
+            found = True
+            break
+        time.sleep(1)
+        print("Waiting for the file in s3")
+
+    if not found:
+        raise Exception("File not found in s3")
+
+    do_assert(len(response["Contents"]) > 0)
+    key = response["Contents"][0]["Key"]
+
+    # Example key:
+    # autoscaler-agent/year=2024/month=04/day=12/07:19:27Z_DGFMM7kfyLFdJASCnJfEYT.ndjson.gz
+    do_assert(key.startswith("autoscaler-agent"))
+    do_assert(key.endswith(".ndjson.gz"))
+
+    response = s3.get_object(Bucket="metrics", Key=key)
+    result = subprocess.check_output(
+        ["gzip", "-d", "-c"], input=response["Body"].read()
+    )
+    result = json.loads(result)
+
+    do_assert("events" in result)
+
+    events = result["events"]
+    do_assert(len(events) > 0, "No events found")
+
+    # Example events[0]:
+    #
+    # {
+    #   "idempotency_key": "2024-04-12T12:10:00.845237Z-autoscaler-agent-nwcnd-1/4",
+    #   "metric": "effective_compute_seconds",
+    #   "type": "incremental",
+    #   "endpoint_id": "test",
+    #   "start_time": "2024-04-12T12:09:55.844007537Z",
+    #   "stop_time": "2024-04-12T12:10:00.845237402Z",
+    #   "value": 1
+    # }
+    do_assert(set(events[0].keys()) == set(
+        [
+            "idempotency_key",
+            "metric",
+            "type",
+            "endpoint_id",
+            "start_time",
+            "stop_time",
+            "value",
+        ]
+    ), "actual keys: %s" % events[0].keys())
+
+    start_time = datetime.datetime.strptime(
+        events[0]["start_time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
+    )
+    stop_time = datetime.datetime.strptime(
+        events[0]["stop_time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
+    )
+
+    # We have a 5 second interval, 10 seconds is a reasonable upper bound
+    do_assert(stop_time - start_time < datetime.timedelta(seconds=10)) 
+    
+    now = datetime.datetime.utcnow()
+    do_assert(now - stop_time < datetime.timedelta(seconds=10))
+
+
+
+class KubctlForward:
+    def __init__(self):
+        self.namespace = os.environ.get("NAMESPACE")
+        self.port = 56632  # arbitrary port
+
+    def start(self):
+        self.process = subprocess.Popen(
+            [
+                "kubectl",
+                "port-forward",
+                "service/minio-service",
+                "%s:9000" % self.port,
+                "-n",
+                self.namespace,
+            ]
+        )
+
+        # Check if the process is still running after 1 second.  
+        # If the port is already in use, it would have exited.
+        time.sleep(1)
```

Comment:
Anecdotally, I've seen port-forwarding take a while to set up — I'd be worried about this `sleep(1)` introducing some flakiness to our e2e tests.

If you haven't seen any issues with it though, then rate of occurrence is probably low enough that it's ok for now.